taichi/transforms/make_cpu_multithreaded_range_for.cpp

```diff
@@ -0,0 +1,126 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+#include "taichi/transforms/utils.h"
+
+namespace taichi::lang {
+
+namespace {
+
+using TaskType = OffloadedStmt::TaskType;
+
+class MakeCPUMultithreadedRangeFor : public BasicStmtVisitor {
+ public:
+  MakeCPUMultithreadedRangeFor(const CompileConfig &config)
+      : config(config), modified(false) {
+  }
+
+  void visit(Block *block) override {
+    for (auto &s_ : block->statements) {
+      s_->accept(this);
+    }
+  }
+
+  void visit(OffloadedStmt *offloaded) override {
+    if (offloaded->task_type != TaskType::range_for) {
+      return;
+    }
+
+    auto offloaded_body = std::make_unique<Block>();
+    auto one = offloaded_body->insert(
+        Stmt::make_typed<ConstStmt>(TypedConstant(PrimitiveType::i32, 1)));
+    auto minimal_block_range = offloaded_body->insert(
+        Stmt::make_typed<ConstStmt>(TypedConstant(PrimitiveType::i32, 512)));
+    auto num_threads = offloaded_body->insert(Stmt::make_typed<ConstStmt>(
+        TypedConstant(PrimitiveType::i32, config.cpu_max_num_threads)));
+    auto thread_index =
+        offloaded_body->insert(Stmt::make_typed<LoopIndexStmt>(offloaded, 0));
+
+    // Retrieve range-for bounds.
+    Stmt *begin_stmt;
+    Stmt *end_stmt;
+    if (offloaded->const_begin) {
+      begin_stmt = offloaded_body->insert(Stmt::make_typed<ConstStmt>(
+          TypedConstant(PrimitiveType::i32, offloaded->begin_value)));
+    } else {
+      begin_stmt = offloaded_body->insert(Stmt::make<GlobalTemporaryStmt>(
+          offloaded->begin_offset, PrimitiveType::i32));
+      begin_stmt =
+          offloaded_body->insert(Stmt::make<GlobalLoadStmt>(begin_stmt));
+    }
+    if (offloaded->const_end) {
+      end_stmt = offloaded_body->insert(Stmt::make_typed<ConstStmt>(
+          TypedConstant(PrimitiveType::i32, offloaded->end_value)));
+    } else {
+      end_stmt = offloaded_body->insert(Stmt::make<GlobalTemporaryStmt>(
+          offloaded->end_offset, PrimitiveType::i32));
+      end_stmt = offloaded_body->insert(Stmt::make<GlobalLoadStmt>(end_stmt));
+    }
+
+    // Inner serial block range is
+    // $[max(((end - begin) + (num_threads - 1)) / num_threads, 1)]
```

Comment:
```suggestion
    // max(((end - begin) + (num_threads - 1)) / num_threads, 1)
```

Reply:
Ah exactly!  The max was added after some bug fixing. Thanks!